toolchain/sem_ir/inst_namer.cpp

```diff
@@ -355,27 +356,45 @@ auto InstNamer::AddBlockLabel(ScopeId scope_id, SemIR::LocId loc_id,
   AddBlockLabel(scope_id, branch.target_id, name.str(), loc_id);
 }
 
-// TODO: Consider addressing recursion here. It may be important because
-// InstNamer is used for debug info.
-// NOLINTNEXTLINE(misc-no-recursion)
 auto InstNamer::CollectNamesInBlock(ScopeId scope_id, InstBlockId block_id)
     -> void {
   if (block_id.is_valid()) {
     CollectNamesInBlock(scope_id, sem_ir_->inst_blocks().Get(block_id));
   }
 }
 
-// NOLINTNEXTLINE(misc-no-recursion): See above TODO.
-auto InstNamer::CollectNamesInBlock(ScopeId scope_id,
+auto InstNamer::CollectNamesInBlock(ScopeId top_scope_id,
                                     llvm::ArrayRef<InstId> block) -> void {
-  Scope& scope = GetScopeInfo(scope_id);
+  std::deque<std::pair<ScopeId, InstId>> insts;
+
+  // Appends a scope and instructions to walk. Avoids recursion while allowing
+  // the loop to below append more instructions during iteration.
+  auto append_block_id = [&](ScopeId scope_id, InstBlockId block_id) {
+    if (block_id.is_valid()) {
+      for (auto inst_id : sem_ir_->inst_blocks().Get(block_id)) {
+        insts.emplace_back(scope_id, inst_id);
+      }
+    }
+  };
+  auto append_block_insts = [&](ScopeId scope_id,
+                                llvm::ArrayRef<InstId> inst_ids) {
+    for (auto inst_id : inst_ids) {
+      insts.emplace_back(scope_id, inst_id);
+    }
+  };
+
+  append_block_insts(top_scope_id, block);
 
   // Use bound names where available. Otherwise, assign a backup name.
-  for (auto inst_id : block) {
+  while (!insts.empty()) {
+    auto [scope_id, inst_id] = insts.front();
+    insts.pop_front();
```

Comment:
```suggestion
    auto [scope_id, inst_id] = insts.pop_back_val();
```

(part of the SmallVector change)